### PR TITLE
Add new unit tests and some refactoring to make testing easier

### PR DESCRIPTION
**Pull request by extramileit**
- Extract TransactionRequestDtoFactory from CardanoTokenSender
- Extract EligibleDonationsService from DonationJob
- Use @Inject instead of constructor dependencies in various places
- Add BlockChainServicesFactory to build all service beans related to BlockFrost API

### CODE REVIEW

**Comment by extramileit** [comment]
Hi Matti
Please feel free to ignore this PR, if you like.  Your original code is well designed and high quality!  The refactoring you see in the PR has mainly to do with making unit testing easier.  My PR ended up being far more invasive than I wanted.  And you won't hurt my feelings by discarding the PR.   I didn't do a comprehensive job writing tests either.  I only wrote a couple, so that I could spend time with the code and learn how to the blockfrost/bloxbean API.

On the other hand, you may find some value in the refactoring, which would be great. I'd also be glad to collaborate with you on other Java project needs.  I've been trying to learn smart contracts using Haskell and more recently trying to use the PAB beta codebase.  That has been rather frustrating, and I feel like I'm not accomplishing much.  Perhaps, I should continue to keep a hand in the Java work, as it appears one can accomplish a lot using Blockfrost/Bloxbean APIs.  Are you working on Bloxbean API, btw?

Thanks
Kurt

**Comment by Cardano-Fans** [comment]
Kurt... I will have a look, please join Cardano Fans on telegram: https://t.me/cardano_fans

Also reminder, on Fridays, 13.15 - 14.15 UTC we have special ZOOM / Google Meet session for developers.

**Comment by extramileit** [comment]
Hi, thanks for the invite to your Telegram group and the developer meeting.  I added myself to the cardano_fans telegram group.  Unfortunately, I wasn't paying close attention to the timezone of the DEV meeting, so I missed that.  I should be able to make the next one, though. :)
Thanks

File: src/main/java/crfa/app/infrastructure/DonationAccountFactory.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    public String readWalletPassFile(String walletPassFile) {
+        try {
+            var file = new File(FileUtils.getUserDirectory(), walletPassFile);
+            log.info("Reading wallet from file: {}", file);
+
+            return FileUtils.readFileToString(file, UTF_8);
+        } catch (IOException e) {
+            var msg = "Unable to read: " + walletPassFile;
+            log.error(msg, e);
+            System.exit(-1);
```
Comment (andreAmorimF) [inline src/main/java/crfa/app/infrastructure/DonationAccountFactory.java]: do the exception will be thrown after the call on `System.exit` ?

File: src/main/java/crfa/app/service/transaction/TransactionRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    @Getter
+    private final Metadata metadata;
+
+    public TransactionRequestDto(ImmutableList<PaymentTransaction> paymentTransactions,
+                                 TransactionDetailsParams transactionDetailsParams, Metadata metadata) {
+        this.paymentTransactions = paymentTransactions;
+        this.transactionDetailsParams = transactionDetailsParams;
+        this.metadata = metadata;
+    }
+}
```
Comment (andreAmorimF) [inline src/main/java/crfa/app/service/transaction/TransactionRequestDto.java]: add newline

File: build.gradle [hunk 3]
```diff
@@ -65,3 +71,7 @@ java {
     sourceCompatibility = JavaVersion.toVersion("11")
     targetCompatibility = JavaVersion.toVersion("11")
 }
+
+test {
+    useJUnitPlatform()
+}
```
Comment (andreAmorimF) [inline build.gradle]: add newline

File: src/main/java/crfa/app/service/transaction/TransactionRequestDtoFactory.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        memos.forEach(cborMemos::add);
+
+        var memo = new CBORMetadataMap()
+                .put(METADATA_MESSAGES_KEY, cborMemos);
+
+        return new CBORMetadata()
+                .put(new BigInteger(CIP20_META_DATUM_LABEL), memo);
+    }
+
+}
```
Comment (andreAmorimF) [inline src/main/java/crfa/app/service/transaction/TransactionRequestDtoFactory.java]: add newline

File: src/test/java/crfa/app/service/CardanoTokenSenderTest.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        Result<TransactionResult> transactionResultResult = tokenSender.sendDonations(donations, sentBy);
+        TransactionResult value = transactionResultResult.getValue();
+        assertSame(txResult, value);
+    }
+
+    private Donation createDonation(String address, BigInteger amount) {
+        return new Donation("1", 1, 1, null, Cadence.EPOCH, "entity", address, amount, null);
+    }
+
+}
```
Comment (andreAmorimF) [inline src/test/java/crfa/app/service/CardanoTokenSenderTest.java]: add newline

File: src/test/java/crfa/app/service/CardanoTokenSenderTest.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.Collections;
+import java.util.List;
+
+import static crfa.app.service.CardanoTokenSender.HEREAFTER_SLOT_COUNT;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.mockito.Mockito.when;
+
+@SuppressWarnings("unchecked")
+@ExtendWith(MockitoExtension.class)
+class CardanoTokenSenderTest {
```
Comment (andreAmorimF) [inline src/test/java/crfa/app/service/CardanoTokenSenderTest.java:38]: 👍🏼

**Review by andreAmorimF — APPROVED** [review]

File: src/main/java/crfa/app/infrastructure/DonationAccountFactory.java (hunk 1 above)
Comment (extramileit) [reply src/main/java/crfa/app/infrastructure/DonationAccountFactory.java]: Nice catch. I got rid of the System.exit and just rely on the bean instantiation to fail, which prevents application from starting up.

File: src/main/java/crfa/app/service/transaction/TransactionRequestDto.java (hunk 2 above)
Comment (extramileit) [reply src/main/java/crfa/app/service/transaction/TransactionRequestDto.java]: done

File: build.gradle (hunk 3 above)
Comment (extramileit) [reply build.gradle]: done

File: src/test/java/crfa/app/service/CardanoTokenSenderTest.java (hunk 5 above)
Comment (extramileit) [reply src/test/java/crfa/app/service/CardanoTokenSenderTest.java]: done

**Comment by extramileit** [comment]
Thanks for the review.  Fixed changes.

**Comment by matiwinnetou** [comment]
Thank You very much!

**Pull request merged**